### DISCUSSION TITLE
Dagster code location - large Docker image size in Kubernetes

### DISCUSSION THREAD

**Discussion by stevenmurphy12** [dagster-io/dagster#17634] [Q&A]
I'm running an OSS deployment of Dagster on Kubernetes, testing assets on an ad-hoc basis at the moment. I noticed that my Docker image is around 700mb, I reduced it to that from 900mb by removing some un-needed PyPi dependencies. 

There aren't any problems at the moment, though my concern is that as we ramp up our usage of Dagster and have potentially many jobs/pods running in parallel, it'll rapidly eat up the memory on the K8s cluster.

I haven't looked into reducing it further yet, though I'm curious: has anyone encountered this before? Is that a reasonable image size? My gut feel is that it's a very large image, though I'm open to ideas.

**Comment by stevenmurphy12** [comment]
For posterity, I reduced the image size with the following steps:
- 887mb - Starting point
- 724mb - Remove pip cache from image as it's not required (`RUN pip install -r requirements_docker_build.txt --no-cache-dir`)
- 347mb - Remove `dagster-gcp` package. This was pulling in a lot of large transitive dependencies which I don't require. I only required one class from this package (`BigQueryResource`), which I copied into my own repo.

**Comment by abkfenris** [comment]
We've been using [Micromamba](https://hub.docker.com/r/mambaorg/micromamba) as our base image to help keep sizes down, and setting the Conda cache directory as a run mount to exclude that from the image, but use it when available to speed up building.

```Dockerfile
RUN --mount=type=cache,id=dagster_code_cache,target=/opt/conda/pkgs,uid=${NEW_MAMBA_USER_ID},gid=${NEW_MAMBA_USER_GID} \
    micromamba install -y -n base -f /tmp/environment.yml
```

Most of our images are ending up 2-400 MB compressed with a good helping of geospatial libraries.
